examples/parser/Parser/Core.roc

```diff
@@ -0,0 +1,292 @@
+
+# -- Generic parsers:
+## Most general way of running a parser.
+##
+## Can be tought of turning the recipe of a parser into its actual parsing function
+## and running this function on the given input.
+##
+## Many (but not all!) parsers consume part of `input` when they succeed.
+## This allows you to string parsers together that run one after the other:
+## The part of the input that the first parser did not consume, is used by the next parser.
+## This is why a parser returns on success both the resulting value and the leftover part of the input.
+##
+## Of course, this is mostly useful when creating your own internal parsing building blocks.
+## `run` or `Parser.Str.runStr` etc. are more useful in daily usage.
+parsePartial : Parser input a, input -> ParseResult input a
+parsePartial = \@Parser parser, input ->
+    parser input
+
+## Runs a parser on the given input, expecting it to fully consume the input
+##
+## The `input -> Bool` parameter is used to check whether parsing has 'completed',
+## (in other words: Whether all of the input has been consumed.)
+##
+## For most (but not all!) input types, a parsing run that leaves some unparsed input behind
+## should be considered an error.
+parse : Parser input a, input, (input -> Bool) -> Result a [ParsingFailure Str, ParsingIncomplete input]
+parse = \parser, input, isParsingCompleted ->
+    when parsePartial parser input is
+        Ok { val: val, input: leftover } ->
+            if isParsingCompleted leftover then
+                Ok val
+            else
+                Err (ParsingIncomplete leftover)
+
+        Err (ParsingFailure msg) ->
+            Err (ParsingFailure msg)
+
+## Parser that can never succeed, regardless of the given input.
+## It will always fail with the given error message.
+##
+## This is mostly useful as 'base case' if all other parsers
+## in a `oneOf` or `alt` have failed, to provide some more descriptive error message.
+fail : Str -> Parser * *
+fail = \msg ->
+    buildPrimitiveParser \_input -> Err (ParsingFailure msg)
+
+## Parser that will always produce the given `val`, without looking at the actual input.
+## This is useful as basic building block, especially in combination with
+## `map` and `apply`.
+const : a -> Parser * a
+const = \val ->
+    buildPrimitiveParser \input ->
+        Ok { val: val, input: input }
+
+## Try the `first` parser and (only) if it fails, try the `second` parser as fallback.
+alt : Parser input a, Parser input a -> Parser input a
+alt = \first, second ->
+    buildPrimitiveParser \input ->
+        when parsePartial first input is
+            Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+            Err (ParsingFailure firstErr) ->
+                when parsePartial second input is
+                    Ok { val: val, input: rest } -> Ok { val: val, input: rest }
+                    Err (ParsingFailure secondErr) ->
+                        Err (ParsingFailure ("\(firstErr) or \(secondErr)"))
```

Comment:
This might read easier as
```suggestion
    buildPrimitiveParser \input ->
       parsePartial first input
       |> Result.onErr (\ParsingFailure firstErr ->
           parsePartial second input
           |> Result.mapErr (\ParsingFailure secondErr ->
               ParsingFailure "\(firstErr) or \(secondErr)"))
```